crates/shim/src/io.rs

```diff
@@ -0,0 +1,299 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+extern crate libc;
+
+use std::fmt::Debug;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::fs::OpenOptionsExt;
+use std::process::Command;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+use crossbeam::sync::WaitGroup;
+use nix::sys::termios::Termios;
+
+use crate::error::{Error, Result};
+use crate::util::IntoOption;
+
+pub trait WriteCloser: Write {
+    fn close(&self);
+}
+
+pub trait IO: Sync + Send {
+    /// Return write side of stdin
+    fn stdin(&self) -> Option<Box<dyn WriteCloser + Send + Sync>>;
+
+    /// Return read side of stdout
+    fn stdout(&self) -> Option<Box<dyn Read + Send>>;
+
+    /// Return read side of stderr
+    fn stderr(&self) -> Option<Box<dyn Read + Send>>;
+
+    /// Set IO for passed command.
+    /// Read side of stdin, write side of stdout and write side of stderr should be provided to command.
+    fn set(&self, cmd: &mut Command) -> Result<()>;
+
+    /// Only close write side (should be stdout/err "from" runc process)
+    fn close_after_start(&self);
+}
+
+pub struct NullIO {}
+
+impl IO for NullIO {
+    fn stdin(&self) -> Option<Box<dyn WriteCloser + Send + Sync>> {
+        None
+    }
+
+    fn stdout(&self) -> Option<Box<dyn Read + Send>> {
+        None
+    }
+
+    fn stderr(&self) -> Option<Box<dyn Read + Send>> {
+        None
+    }
+
+    fn set(&self, cmd: &mut Command) -> Result<()> {
+        cmd.stdout(std::process::Stdio::null());
+        cmd.stderr(std::process::Stdio::null());
+        Ok(())
+    }
+
+    fn close_after_start(&self) {}
+}
+
+pub struct FIFO {
+    pub stdin: Option<String>,
+    pub stdout: Option<String>,
+    pub stderr: Option<String>,
+}
+
+impl IO for FIFO {
+    fn stdin(&self) -> Option<Box<dyn WriteCloser + Send + Sync>> {
+        None
+    }
+
+    fn stdout(&self) -> Option<Box<dyn Read + Send>> {
+        None
+    }
+
+    fn stderr(&self) -> Option<Box<dyn Read + Send>> {
```

Comment:
Should we just return `None` from the `Io` trait as default implementaiton and here just implement whats needs to be implemented?